source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+        {
+          Particle<dim> p(pdata,property_manager->get_particle_size());
+
+          if (status == parallel::distributed::Triangulation<dim>::CELL_COARSEN
+              || status == parallel::distributed::Triangulation<dim>::CELL_PERSIST)
+            particles.insert(std::make_pair(std::make_pair(cell->level(),cell->index()),p));
+          else if (status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+            {
+              for (unsigned int child_index = 0; child_index < cell->number_of_children(); ++child_index)
+                {
+                  const typename parallel::distributed::Triangulation<dim>::cell_iterator child = cell->child(child_index);
+                  try
+                    {
+                      const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(child, p.get_location());
+                      if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                        {
+                          particles.insert(std::make_pair(std::make_pair(child->level(),child->index()),p));
+                        }
+                    }
+                  catch (...)
+                    {}
+                }
+            }
+        }
+
+      if (particle_load_balancing == remove_and_add_particles)
+        {
+          // TODO: Add particles
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::sort_particles_in_subdomains_and_cells()
+    {
+      // There are three reasons why a particle is not in its old cell:
+      // It moved to another cell, to another domain or it left the mesh.
+      // Sort the particles accordingly and deal with them
+      std::multimap<types::LevelInd, Particle<dim> >            moved_particles_cell;
+      std::multimap<types::subdomain_id, Particle<dim> > moved_particles_domain;
+      std::multimap<types::LevelInd, Particle<dim> >            lost_particles;
+
+      // Find the cells that the particles moved to.
+      // Note that the iterator in the following loop is increased in a
+      // very particular way, because it is changed, if elements
+      // get erased. A change can result in invalid memory access.
+      typename std::multimap<types::LevelInd, Particle<dim> >::iterator   it;
+      for (it=particles.begin(); it!=particles.end();)
+        {
+          // If we know the particles old cell, check if it is still inside
+          if (it->first != std::make_pair(-1,-1))
+            {
+              const typename parallel::distributed::Triangulation<dim>::active_cell_iterator
+              old_cell (&(this->get_triangulation()), it->first.first, it->first.second);
+
+              try
+                {
+                  const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(old_cell, it->second.get_location());
+                  if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                    {
+                      ++it;
+                      continue;
```

Comment:
leave a comment saying that the particle is still inside, so move on to the next cell